### PR TITLE
Parser and Transpiler for JSX like syntax

### PR DESCRIPTION
**Pull request by jonahwilliams**

### CODE REVIEW

File: tool/example.dart [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import 'package:butterfly/butterfly.dart';
+
+class FooWidget extends StatelessWidget {
+  @override
+  Node build() {
+    var local = false;
+    var bar = 2;
+    var items = [1, 2, 3].map((x) => text("${x}")).toList();
+
+    return new MyComponent(foo: bar)([
```
Comment (yjbanov) [inline tool/example.dart]: Not important right now, but I wonder if we could generate the code in a way that doesn't shift line numbers.

File: tool/parser.dart [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+  Parser();
+
+  int get peek => _runes[_index];
+
+  /// Parses the given dart extension (dx) fragment into dart code.
+  Ast parse(String source) {
+    _nodes = <Ast>[];
+    _runes = source.runes.toList();
+    _index = 0;
+    consumeWhitespace();
```
Comment (yjbanov) [inline tool/parser.dart]: Do you need to `consumeWhitespace()` here if the first thing inside `while(true)` already does it?

File: tool/example.dx.dart [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+class FooWidget extends StatelessWidget {
+  @override
+  Node build() {
+    var local = false;
+    var bar = 2;
+    var items = [1, 2, 3].map((x) => Dx('''"${x}"''')).toList();
+
+    return Dx('''
+      <MyComponent foo=${bar}>
+        "This is text"
```
Comment (yjbanov) [inline tool/example.dx.dart]: I'm tempted to suggest that we introduce a `<text>` tag and end the "how should we handle whitespace between tags" discussion :)

Comment (yjbanov) [inline tool/example.dx.dart]: Same with self-closing tags.

File: tool/parser.dart [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+  void consume([int chars = 1]) {
+    _index += chars;
+  }
+
+  bool isDone() => _index >= _runes.length;
+
+  static bool isLetter(int char) =>
+      (char >= 0x41 && char <= 0x5A) || (char >= 0x61 && char <= 0x7A);
+
+  static bool isNum(int char) => (char >= 0x30 && char <= 0x39);
```
Comment (yjbanov) [inline tool/parser.dart]: nit: `isDigit`?

File: tool/transpiler.dart [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import 'dart:io';
+import 'parser.dart';
+
+/// This is a super hack, a real transpiler would at least use the analysis api to grab to string
```
Comment (yjbanov) [inline tool/transpiler.dart]: Let's mark hacks with `TODO`

File: tool/parser.dart (hunk 4 above)
Comment (jonahwilliams) [inline tool/parser.dart]: Ahh, that's what those are called.

Done.

File: tool/example.dx.dart (hunk 3 above)
Comment (jonahwilliams) [inline tool/example.dx.dart]: Lets come back to text nodes, there are definitely a few different ways we can solve it.  For self closing nodes I am treating them just as 'I have no children'

File: tool/transpiler.dart (hunk 5 above)
Comment (jonahwilliams) [inline tool/transpiler.dart]: Done.

File: tool/parser.dart (hunk 2 above)
Comment (jonahwilliams) [inline tool/parser.dart]: Nope, good catch

File: tool/example.dart (hunk 1 above)
Comment (jonahwilliams) [inline tool/example.dart]: Sounds like a challenge 🥇

File: tool/example.dx.dart (hunk 3 above)
Comment (yjbanov) [inline tool/example.dx.dart]: Not important at this stage.

**Pull request merged**